students/Tianx/Lesson5/mailroom3.py

```diff
@@ -0,0 +1,189 @@
+# ------------------------------------------#
+# !/usr/bin/env python3
+# Title: mailroom3.py
+# Desc: Mailroom Part 3
+# Tian Xie, 2020-04-12, Created File
+# Tian Xie, 2020-04-26, Adding exception handler
+# ------------------------------------------#
+
+# A dictionary of donor paired with a list of donation amounts
+dict_of_donors = {'Jeff Bezos': [1.00, 50.00],
+                  'Warren Buffet': [100.00, 1000.00],
+                  'Bill Gates': [100.00, 500.00],
+                  'Tim Cook': [300.00],
+                  'Jack Ma': [2000.00]}
+# Main Prompt
+main_prompt = "\n".join(("=======  Main Menu ======= \n",
+                         "Welcome to the Mailroom. Please select an option:",
+                         "1: Send a Thank You to a single donor",
+                         "2: Create a Report",
+                         "3: Send letters to all donors",
+                         "4: Exit",
+                         ">>> "))
+
+
+def menu_selection(prompt, dispatch_dict):
+    """Displays a menu of choices to the user
+    Args:
+        prompt: main menu
+        dispatch_dict: choices for user to choose from
+    Returns:
+        None.
+    """
+    while True:
+        response = input(prompt)  # continuously collect user selection
+        while True:
+            try:
+                int(response)
+                break
+            except ValueError:
+                response = input('Error: Please enter a number. Please select again >')
+        if dispatch_dict[response]() == "Exit Menu":
+            break
+
+
+def show_donor_list(donor_list):
+    """Displays current donor list.
+
+    Args:
+        donor_list: 2D data structure (dictionary of objects).
+
+    Returns:
+        None.
+
+    """
+    print('======= The Donor List: =======')
+    print('Donor Name:\n')
+    # Not using comprehension here.
+    for donor in donor_list:
+        print(donor)
+    print('======================================')
+
+
+def adding_donor_info(name, donation, donor_list):
+    """Adding donor info to the list.
+
+    Args:
+        name: donor name
+        donation: donation amount
+        donor_list: 2D data structure (dictionary of objects).
+
+    Returns:
+        None.
+
+    """
+    if name not in donor_list:
+        added_donor = {name: [donation]}
+        donor_list.update(added_donor)
+    else:
+        donor_list[name].insert(0, donation) # If donor name exists, insert the donation amount rather than updating the dictionary
```

Comment:
a more common pattern would be to append rather than insert, insert would indicate that there is a specific custom logic for this amount to be at the beginning of the list.

`donors[name].append(donation)`